mingw-w64-git/PKGBUILD

```
@@ -0,0 +1,563 @@
+# Maintainer: Johannes Schindelin <johannes.schindelin@gmx.de>
+
+_realname=git
+pkgbase="mingw-w64-${_realname}"
+pkgname=("${MINGW_PACKAGE_PREFIX}-${_realname}"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-doc-html"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-doc-man"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-subtree"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-credential-wincred"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-perl"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-send-email"
+	 "${MINGW_PACKAGE_PREFIX}-gitweb"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-svn"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-archimport"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-cvs"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-p4"
+	 "${MINGW_PACKAGE_PREFIX}-gitk"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-gui"
+	 "${MINGW_PACKAGE_PREFIX}-${_realname}-for-windows-addons")
+tag=2.52.0.windows.1
+pkgver=2.52.0.1
+pkgrel=1
+pkgdesc="The fast distributed version control system (mingw-w64)"
+arch=('any')
```

```suggestion
arch=('any')
mingw_arch=('mingw32' 'mingw64' 'ucrt64' 'clang64' 'clangarm64')
```

---

```suggestion
arch=('any')
mingw_arch=('mingw32' 'mingw64' 'clangarm64')
```

Since we would still need some changes to git-for-windows/git for proper `UCRT64` and `CLANG64` support we should exclude those envs for now.

---

> > what changes do you have in mind for `compat/mingw.c` and `environment.c`?
> 
> I see these:
> - `compat/mingw.c`:
>   - [`append_system_bin_dirs()`](https://github.com/git-for-windows/git/blob/v2.52.0.windows.1/compat/mingw.c#L3646-L3664) and [the `MSYSTEM` definition in `setup_windows_environment()`](https://github.com/git-for-windows/git/blob/v2.52.0.windows.1/compat/mingw.c#L3760-L3767); Both are only in effect if `MSYSTEM` is unset, which should be never in MSYS2. But yes, it would be nice to fix this in a generic way, passing the relevant values as string constants to the build rather than hard-coding them based on data type sizes and CPU macros.

Those are indeed what I found on a quick search,  but I hadn't realized these where less pressing in Msys2 than Git for Windows.

> - `environment.c`:
>   - There are [stack size work-arounds](https://github.com/git-for-windows/git/blob/v2.52.0.windows.1/environment.c#L94-L103) special-casing `clangarm64`. Apart from noticing that there is now a duplicated clause which we need to drop from Git for Windows ([it's not in upstream Git](https://github.com/git/git/blob/v2.52.0/environment.c#L84-L106)) I am not quite certain whether we have to do something about it right now. I guess we should investigate whether the same [exit code 127 lets t6700 fail](https://github.com/git-for-windows/git/commit/436a42215e51fa2f8b74d128472d7d9bfe2595e1), and adjust the constant accordingly if necessary.

The fact that we have special cases for MSVC and CLANGARM64 makes me a little suspicious that this problem might be less arm specific than we hope.

---

> `environment.c`:
> 
>     * There are [stack size work-arounds](https://github.com/git-for-windows/git/blob/v2.52.0.windows.1/environment.c#L94-L103) special-casing `clangarm64`.

@rimrul turns out that your worries were [spot on](https://github.com/msys2/MINGW-packages/actions/runs/19572657970/job/56049729289?pr=26470#step:9:15137) (the link takes a while to jump to the correct line in the log, [if](https://www.mistys-internet.website/blog/blog/2024/07/12/github-is-starting-to-feel-like-legacy-software/) it jumps there at all):

```
[...]
2025-11-21T14:05:53.2159579Z expecting success of 6700.4 'default limit for git-archive fails gracefully': 
2025-11-21T14:05:53.2160050Z 	test_must_fail git archive big >/dev/null
2025-11-21T14:05:53.2160239Z 
2025-11-21T14:05:53.2162328Z ++ test_must_fail git archive big
2025-11-21T14:05:53.2162717Z ++ case "$1" in
2025-11-21T14:05:53.2162926Z ++ _test_ok=
2025-11-21T14:05:53.2163222Z ++ test_must_fail_acceptable git archive big
2025-11-21T14:05:53.2163579Z ++ test git = env
2025-11-21T14:05:53.2163834Z ++ test git = nongit
2025-11-21T14:05:53.2164103Z ++ case "$1" in
2025-11-21T14:05:53.2164352Z ++ return 0
2025-11-21T14:05:53.2164579Z ++ git archive big
2025-11-21T14:05:53.2601791Z ++ exit_code=127
2025-11-21T14:05:53.2602181Z ++ test 127 -eq 0
2025-11-21T14:05:53.2602470Z ++ test_match_signal 13 127
2025-11-21T14:05:53.2602793Z ++ test 127 = 141
2025-11-21T14:05:53.2603058Z ++ test 127 = 269
2025-11-21T14:05:53.2603304Z ++ return 1
2025-11-21T14:05:53.2603529Z ++ test 127 -gt 129
2025-11-21T14:05:53.2603788Z ++ test 127 -eq 127
2025-11-21T14:05:53.2604186Z ++ echo 'test_must_fail: command not found: git archive big'
2025-11-21T14:05:53.2604733Z test_must_fail: command not found: git archive big
2025-11-21T14:05:53.2605147Z ++ return 1
2025-11-21T14:05:53.2606476Z error: last command exited with $?=1
2025-11-21T14:05:53.2608706Z not ok 4 - default limit for git-archive fails gracefully
2025-11-21T14:05:53.2816610Z #	
2025-11-21T14:05:53.2817243Z #		test_must_fail git archive big >/dev/null
2025-11-21T14:05:53.2817869Z #	
2025-11-21T14:05:53.2938027Z 1..4
2025-11-21T14:05:53.3417741Z ==> ERROR: A failure occurred in build().
2025-11-21T14:05:53.3647555Z     Aborting...
[...]
```

So the work-around is not specific to clang/ARM64, it seems specific to `clang` on Windows. (I know it's not clang on x86_64 in general because the Linux jobs of Git's CI pass the test suite just fine).

---

@500-internal-server-error you do realize, don't you, that Git _requires_ a Unix shell interpreter to do almost everything, from opening an editor to running hooks to evaluating shell aliases? It's even required to spawn `receive-pack` in a clone. So yeah, Git is very much not independent of the Unix emulation layer, so there.

---

@rimrul I had no idea that you were also working on this, sorry for stepping on your toes!

That's good though because you can now review my patches and see whether there are any gaps 😄  Do you see anything that's missing from my branch?